File: leetcode/129_sum_root_to_leaf.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        string = ''
        self.sum = 0
        self.dfs(root, string)
        return self.sum
    
    def dfs(self, root, string):
        if not root:
            return
            
        if not root.left and not root.right:
            string += str(root.val)
            print(string)
            self.sum += int(string)
            return
        
        string += str(root.val)

        self.dfs(root.left, string)
        self.dfs(root.right, string)

File: leetcode/129_sum_root_to_leaf.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    vector<int> res;
public:
    int sumNumbers(TreeNode* root) {
        dfs(root, "");
        int sum = 0;
        for (const auto& num : res) {
            cout << num << '\n';
            sum += num;
        }

        return sum;
    }

    void dfs(TreeNode* node, string num) {
        if (node == nullptr)
            return;
        if (node->left == nullptr && node->right == nullptr) {
            num += to_string(node->val);
            res.push_back(stoi(num));
            return;
        }

        num += to_string(node->val);
        dfs(node->left, num);
        dfs(node->right, num);
    }
};
